File: core/tests/test_position_hints.cpp
// Catch2 unit tests for Circuit::set_position / position_hint /
// position_hints — the kernel-side storage added in Phase 2 of
// `add-python-schematic-renderer`.
//
// Hints have no effect on simulation; the kernel just round-trips them
// through the parser → renderer pipeline. Tests cover:
//   - set/get round-trip for both `(layer, slot)` and `(x, y)` forms
//   - snapshot accessor is detached from later mutation
//   - missing-component query returns nullopt
//   - empty-hint rejection at the kernel boundary
//   - determinism across runs (byte-identical state from identical
//     construction sequences)

#include <catch2/catch_test_macros.hpp>

#include "pulsim/v1/runtime_circuit.hpp"

using namespace pulsim::v1;

TEST_CASE("position_hints: empty circuit has no hints",
          "[position_hints]") {
    Circuit ckt;
    REQUIRE(ckt.num_position_hints() == 0);
    REQUIRE(ckt.position_hints().empty());
}

TEST_CASE("position_hints: set/get a (layer, slot) hint",
          "[position_hints]") {
    Circuit ckt;
    ckt.set_position("Q1", /*layer=*/2, /*slot=*/1);

    REQUIRE(ckt.num_position_hints() == 1);
    auto hint = ckt.position_hint("Q1");
    REQUIRE(hint.has_value());
    REQUIRE(hint->layer.has_value());
    REQUIRE(*hint->layer == 2);
    REQUIRE(hint->slot.has_value());
    REQUIRE(*hint->slot == 1);
    REQUIRE_FALSE(hint->x.has_value());
    REQUIRE_FALSE(hint->y.has_value());
}

TEST_CASE("position_hints: set/get an absolute (x, y) hint",
          "[position_hints]") {
    Circuit ckt;
    ckt.set_position("Cout", /*layer=*/std::nullopt, /*slot=*/std::nullopt,
                     /*x=*/200.0, /*y=*/80.0);

    auto hint = ckt.position_hint("Cout");
    REQUIRE(hint.has_value());
    REQUIRE(hint->x.has_value());
    REQUIRE(*hint->x == 200.0);
    REQUIRE(hint->y.has_value());
    REQUIRE(*hint->y == 80.0);
    REQUIRE_FALSE(hint->layer.has_value());
    REQUIRE_FALSE(hint->slot.has_value());
}

TEST_CASE("position_hints: both (layer, slot) AND (x, y) can coexist",
          "[position_hints]") {
    // The kernel persists every set field; the renderer (downstream)
    // decides priority. We just confirm we don't silently drop one.
    Circuit ckt;
    ckt.set_position("M1", 3, 4, 120.0, 60.0);
    auto hint = ckt.position_hint("M1");
    REQUIRE(hint.has_value());
    REQUIRE(*hint->layer == 3);
    REQUIRE(*hint->slot == 4);
    REQUIRE(*hint->x == 120.0);
    REQUIRE(*hint->y == 60.0);
}

TEST_CASE("position_hints: querying an unhinted component returns nullopt",
          "[position_hints]") {
    Circuit ckt;
    ckt.set_position("R1", 0, 0);
    REQUIRE_FALSE(ckt.position_hint("never_set").has_value());
}

TEST_CASE("position_hints: re-setting replaces the previous hint wholesale",
          "[position_hints]") {
    Circuit ckt;
    ckt.set_position("R1", /*layer=*/0, /*slot=*/0);
    ckt.set_position("R1", /*layer=*/std::nullopt, /*slot=*/std::nullopt,
                     /*x=*/50.0, /*y=*/25.0);

    auto hint = ckt.position_hint("R1");
    REQUIRE(hint.has_value());
    // First call's (layer, slot) is gone — replaced, not merged.
    REQUIRE_FALSE(hint->layer.has_value());
    REQUIRE_FALSE(hint->slot.has_value());
    REQUIRE(*hint->x == 50.0);
    REQUIRE(*hint->y == 25.0);
}

TEST_CASE("position_hints: snapshot is detached from later mutation",
          "[position_hints]") {
    Circuit ckt;
    ckt.set_position("R1", 0, 0);
    auto snap = ckt.position_hints();
    REQUIRE(snap.size() == 1);
    REQUIRE(*snap.at("R1").layer == 0);

    ckt.set_position("R1", 5, 5);

    // Snapshot still shows the old values.
    REQUIRE(*snap.at("R1").layer == 0);
    // But the live circuit reflects the new ones.
    REQUIRE(*ckt.position_hint("R1")->layer == 5);
}

TEST_CASE("position_hints: empty hint is rejected",
          "[position_hints]") {
    Circuit ckt;
    REQUIRE_THROWS_AS(
        ckt.set_position("R1"),  // all four optionals default to nullopt
        std::invalid_argument);
    REQUIRE(ckt.num_position_hints() == 0);
}

TEST_CASE("position_hints: hints survive irrelevant Circuit mutations",
          "[position_hints]") {
    // Setting hints before adding devices is allowed (YAML parser
    // pattern). After adding devices the hints must still be queryable.
    Circuit ckt;
    auto a = ckt.add_node("a");
    ckt.set_position("R1", 0, 0);
    ckt.set_position("V1", /*layer=*/std::nullopt, /*slot=*/std::nullopt,
                     /*x=*/10.0, /*y=*/20.0);
    ckt.add_resistor("R1", a, ckt.ground(), 1000.0);
    ckt.add_voltage_source("V1", a, ckt.ground(), 5.0);

    REQUIRE(ckt.position_hint("R1").has_value());
    REQUIRE(*ckt.position_hint("R1")->layer == 0);
    REQUIRE(ckt.position_hint("V1").has_value());
    REQUIRE(*ckt.position_hint("V1")->x == 10.0);
}

TEST_CASE("position_hints: hints have no effect on num_components or topology",
          "[position_hints]") {
    Circuit ckt;
    auto a = ckt.add_node("a");
    ckt.add_resistor("R1", a, ckt.ground(), 100.0);
    const auto before = ckt.num_components();
    ckt.set_position("R1", 0, 0);
    REQUIRE(ckt.num_components() == before);  // hint isn't a device
    // The descriptor list is unchanged by setting a hint.
    auto descs = ckt.components();
    REQUIRE(descs.size() == 1);
    REQUIRE(descs[0].name == "R1");
}

TEST_CASE("position_hints: determinism — identical builds yield identical hints",
          "[position_hints]") {
    auto build = []() {
        Circuit c;
        c.set_position("A", 0, 0);
        c.set_position("B", 1, 0, std::nullopt, std::nullopt);
        c.set_position("C", std::nullopt, std::nullopt, 50.0, 75.0);
        return c;
    };
    auto c1 = build();
    auto c2 = build();
    auto s1 = c1.position_hints();
    auto s2 = c2.position_hints();
    REQUIRE(s1.size() == s2.size());
    for (const auto& [name, hint] : s1) {
        auto it = s2.find(name);
        REQUIRE(it != s2.end());
        REQUIRE(hint.layer == it->second.layer);
        REQUIRE(hint.slot == it->second.slot);
        REQUIRE(hint.x == it->second.x);
        REQUIRE(hint.y == it->second.y);
    }
}

File: legacy/v1/python/tests/test_position_hints.py
"""Phase 2 tests for position-hint storage on Circuit (Python side).

Covers the contract in
`openspec/changes/add-python-schematic-renderer/specs/python-bindings/spec.md`
("Position Hint Binding") plus the YAML round-trip path documented in
`specs/netlist-yaml/spec.md`.

Hints do NOT affect simulation; they're consumed by the renderer in
Phase 3. These tests only verify storage + round-trip behavior.
"""

from __future__ import annotations

from pathlib import Path

import pytest

import pulsim as ps


REPO_ROOT = Path(__file__).resolve().parents[2]


def test_circuit_starts_with_no_hints() -> None:
    ckt = ps.Circuit()
    assert ckt.num_position_hints() == 0
    assert ckt.position_hints() == {}


def test_set_position_layer_slot_round_trips() -> None:
    ckt = ps.Circuit()
    ckt.set_position("Q1", layer=2, slot=1)
    hint = ckt.position_hint("Q1")
    assert hint is not None
    assert hint.layer == 2
    assert hint.slot == 1
    assert hint.x is None
    assert hint.y is None


def test_set_position_xy_round_trips() -> None:
    ckt = ps.Circuit()
    ckt.set_position("Cout", x=200.0, y=80.0)
    hint = ckt.position_hint("Cout")
    assert hint is not None
    assert hint.x == 200.0
    assert hint.y == 80.0
    assert hint.layer is None
    assert hint.slot is None


def test_set_position_both_forms_coexist() -> None:
    ckt = ps.Circuit()
    ckt.set_position("M1", layer=3, slot=4, x=120.0, y=60.0)
    hint = ckt.position_hint("M1")
    assert hint is not None
    assert hint.layer == 3
    assert hint.slot == 4
    assert hint.x == 120.0
    assert hint.y == 60.0


def test_position_hint_missing_returns_none() -> None:
    ckt = ps.Circuit()
    ckt.set_position("R1", layer=0, slot=0)
    assert ckt.position_hint("never_set") is None


def test_position_hints_snapshot_is_detached() -> None:
    ckt = ps.Circuit()
    ckt.set_position("R1", layer=0, slot=0)
    snap = ckt.position_hints()
    assert "R1" in snap
    assert snap["R1"].layer == 0

    ckt.set_position("R1", layer=5, slot=5)

    # The snapshot dict is a copy — mutating the circuit doesn't update it.
    assert snap["R1"].layer == 0
    # But a fresh query reflects the new hint.
    assert ckt.position_hint("R1") is not None
    assert ckt.position_hint("R1").layer == 5  # type: ignore[union-attr]


def test_position_hint_readonly() -> None:
    """`PositionHint` exposes its fields read-only — assignment raises."""
    ckt = ps.Circuit()
    ckt.set_position("Q1", layer=2, slot=1)
    hint = ckt.position_hint("Q1")
    assert hint is not None
    with pytest.raises(AttributeError):
        hint.layer = 99  # type: ignore[misc]


def test_set_position_rejects_empty_hint() -> None:
    """At least one of (layer, slot) / (x, y) MUST be set."""
    ckt = ps.Circuit()
    # pybind11 maps invalid_argument → ValueError by default.
    with pytest.raises(ValueError):
        ckt.set_position("R1")


def test_position_hint_repr_smoke() -> None:
    ckt = ps.Circuit()
    ckt.set_position("Q1", layer=2, slot=1)
    hint = ckt.position_hint("Q1")
    rep = repr(hint)
    assert "PositionHint" in rep
    assert "layer" in rep
    assert "slot" in rep


def test_resetting_overwrites_previous_hint() -> None:
    ckt = ps.Circuit()
    ckt.set_position("R1", layer=0, slot=0)
    ckt.set_position("R1", x=50.0, y=25.0)
    hint = ckt.position_hint("R1")
    assert hint is not None
    # Previous (layer, slot) replaced wholesale — NOT merged.
    assert hint.layer is None
    assert hint.slot is None
    assert hint.x == 50.0
    assert hint.y == 25.0


def test_position_hints_dont_affect_components_list() -> None:
    """Setting a hint never registers a new device."""
    ckt = ps.Circuit()
    a = ckt.add_node("a")
    ckt.add_resistor("R1", a, ckt.ground(), 100.0)
    before = ckt.num_components()
    ckt.set_position("R1", layer=0, slot=0)
    assert ckt.num_components() == before
    # The descriptor list is unchanged.
    descs = ckt.components()
    assert len(descs) == 1
    assert descs[0].name == "R1"


def test_hints_can_be_set_before_devices_are_added() -> None:
    """YAML parser pattern: it reads `position:` and `nodes:` in order
    so the hint may be set before the corresponding `add_*` call."""
    ckt = ps.Circuit()
    ckt.set_position("R1", layer=0, slot=0)
    a = ckt.add_node("a")
    ckt.add_resistor("R1", a, ckt.ground(), 100.0)
    hint = ckt.position_hint("R1")
    assert hint is not None
    assert hint.layer == 0


# ----------------------------------------------------------------------
# YAML round-trip tests (specs/netlist-yaml)
# ----------------------------------------------------------------------


def test_yaml_position_layer_slot_round_trips(tmp_path) -> None:
    """A YAML netlist with `position: {layer, slot}` lands in the
    Circuit's position_hints with the same values."""
    yaml_path = tmp_path / "rc_with_positions.yaml"
    yaml_path.write_text(
        """
schema: pulsim-v1
version: 1
simulation:
  tstop: 1e-3
  dt: 1e-6
components:
  - type: voltage_source
    name: V1
    nodes: [vin, 0]
    waveform: { type: dc, value: 5.0 }
    position: { layer: 0, slot: 0 }
  - type: resistor
    name: R1
    nodes: [vin, vout]
    value: 1k
    position: { layer: 1, slot: 0 }
  - type: capacitor
    name: C1
    nodes: [vout, 0]
    value: 1u
    position: { layer: 2, slot: 0 }
"""
    )
    parser = ps.YamlParser(ps.YamlParserOptions())
    ckt, _ = parser.load(str(yaml_path))

    hints = ckt.position_hints()
    assert set(hints.keys()) == {"V1", "R1", "C1"}
    assert hints["V1"].layer == 0 and hints["V1"].slot == 0
    assert hints["R1"].layer == 1 and hints["R1"].slot == 0
    assert hints["C1"].layer == 2 and hints["C1"].slot == 0


def test_yaml_position_xy_round_trips(tmp_path) -> None:
    yaml_path = tmp_path / "rc_with_xy.yaml"
    yaml_path.write_text(
        """
schema: pulsim-v1
version: 1
simulation:
  tstop: 1e-3
  dt: 1e-6
components:
  - type: voltage_source
    name: V1
    nodes: [vin, 0]
    waveform: { type: dc, value: 5.0 }
    position: { x: 10.0, y: 20.0 }
  - type: resistor
    name: R1
    nodes: [vin, 0]
    value: 100
    position: { x: 150.5, y: 50.5 }
"""
    )
    parser = ps.YamlParser(ps.YamlParserOptions())
    ckt, _ = parser.load(str(yaml_path))
    hints = ckt.position_hints()
    assert hints["V1"].x == 10.0 and hints["V1"].y == 20.0
    assert hints["R1"].x == 150.5 and hints["R1"].y == 50.5


def test_yaml_missing_position_is_not_an_error(tmp_path) -> None:
    """Existing demos (rc / buck / half-bridge) have no `position:`
    field — they MUST still parse cleanly and yield an empty hint set."""
    parser = ps.YamlParser(ps.YamlParserOptions())
    ckt, _ = parser.load(str(REPO_ROOT / "examples" / "buck_converter.yaml"))
    assert ckt.num_position_hints() == 0
    assert ckt.position_hints() == {}


def test_yaml_position_empty_map_warns_and_ignores(tmp_path) -> None:
    """A `position:` map with no recognized keys is a warning, not an
    error. The component is still created without a hint."""
    yaml_path = tmp_path / "rc_bad_position.yaml"
    yaml_path.write_text(
        """
schema: pulsim-v1
version: 1
simulation:
  tstop: 1e-3
  dt: 1e-6
components:
  - type: resistor
    name: R1
    nodes: [a, 0]
    value: 100
    position: { foo: 1, bar: 2 }
"""
    )
    parser = ps.YamlParser(ps.YamlParserOptions())
    ckt, _ = parser.load(str(yaml_path))
    # R1 still created.
    descs = {d.name for d in ckt.components()}
    assert "R1" in descs
    # But no hint set.
    assert ckt.num_position_hints() == 0


def test_yaml_position_mixed_xy_and_layer_slot(tmp_path) -> None:
    """A YAML hint that sets both (layer, slot) and (x, y) on the same
    component keeps all four values."""
    yaml_path = tmp_path / "rc_mixed.yaml"
    yaml_path.write_text(
        """
schema: pulsim-v1
version: 1
simulation:
  tstop: 1e-3
  dt: 1e-6
components:
  - type: resistor
    name: R1
    nodes: [a, 0]
    value: 100
    position: { layer: 3, slot: 4, x: 200.0, y: 80.0 }
"""
    )
    parser = ps.YamlParser(ps.YamlParserOptions())
    ckt, _ = parser.load(str(yaml_path))
    hint = ckt.position_hint("R1")
    assert hint is not None
    assert hint.layer == 3
    assert hint.slot == 4
    assert hint.x == 200.0
    assert hint.y == 80.0
